File: ruby/lib/runapi/gemini_omni/resources/create_character.rb
# frozen_string_literal: true

module RunApi
  module GeminiOmni
    module Resources
      # Builds a reusable character from a reference image and description.
      # Attach audio IDs to give the character a specific voice.
      # Synchronous -- only +run+ is available (no create/get polling).
      class CreateCharacter
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/gemini_omni/create_character"
        RESPONSE_CLASS = Types::CreateCharacterResponse
        DESCRIPTIONS_MAX_LENGTH = 20_000
        CHARACTER_NAME_MAX_LENGTH = 210

        def initialize(http)
          @http = http
        end

        def run(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        private

        def validate_params!(params)
          validate_required!(params, :descriptions)
          validate_required!(params, :reference_image_url)
          validate_array!(params, :audio_ids) if param(params, :audio_ids)
          validate_length!(params, :descriptions, DESCRIPTIONS_MAX_LENGTH)
          validate_length!(params, :character_name, CHARACTER_NAME_MAX_LENGTH)
        end

        def validate_required!(params, key)
          value = param(params, key)
          present = if value.is_a?(Array)
            value.any?
          else
            value.is_a?(String) ? !value.empty? : !value.nil?
          end
          return if present

          raise Core::ValidationError, "#{key} is required"
        end

        def validate_array!(params, key)
          return if param(params, key).is_a?(Array)

          raise Core::ValidationError, "#{key} must be an array"
        end

        def validate_length!(params, key, max_length)
          value = param(params, key)
          return if value.nil? || value.to_s.length <= max_length

          raise Core::ValidationError, "#{key} must be at most #{max_length} characters"
        end
      end
    end
  end
end

File: python/src/runapi/gemini_omni/resources/create_character.py
"""Gemini Omni create-character resource (synchronous)."""

from __future__ import annotations

from typing import Any, Dict

from runapi.core import Resource, ValidationError

from ..contract_gen import CONTRACT
from ..types import CreateCharacterResponse


class CreateCharacter(Resource):
    """Create a reusable character. Synchronous: ``run()`` returns the result directly."""

    ENDPOINT = "/api/v1/gemini_omni/create_character"

    RESPONSE_CLASS = CreateCharacterResponse

    MODEL = "gemini-omni-character"

    DESCRIPTIONS_MAX_LENGTH = 20_000
    CHARACTER_NAME_MAX_LENGTH = 210

    def run(self, **params: Any) -> Any:
        """Create a reusable character (synchronous).

        Args:
            **params: Reusable character parameters (descriptions, reference_image_url, ...).

        Returns:
            The result.
        """
        compacted = self._compact_params(params)
        self._validate_params(compacted)
        return self._request("post", self.ENDPOINT, body=compacted)

    def _validate_params(self, params: Dict[str, Any]) -> None:
        self._validate_contract(CONTRACT["create-character"], {**params, "model": self.MODEL})
        if params.get("audio_ids") is not None:
            self._validate_array(params, "audio_ids")
        self._validate_length(params, "descriptions", self.DESCRIPTIONS_MAX_LENGTH)
        self._validate_length(params, "character_name", self.CHARACTER_NAME_MAX_LENGTH)

    @staticmethod
    def _validate_array(params: Dict[str, Any], key: str) -> None:
        if not isinstance(params.get(key), list):
            raise ValidationError(f"{key} must be an array")

    @staticmethod
    def _validate_length(params: Dict[str, Any], key: str, max_length: int) -> None:
        value = params.get(key)
        if value is None or len(str(value)) <= max_length:
            return
        raise ValidationError(f"{key} must be at most {max_length} characters")
